README.md
# CodeRunner 🏃

> A judge 👨🏽‍⚖️ for your programs, run and test your programs through Python


![PyPI](https://img.shields.io/pypi/v/coderunner?color=blue)
[![Build Status](https://travis-ci.org/codeclassroom/CodeRunner.svg?branch=master)](https://travis-ci.org/codeclassroom/CodeRunner)
[![codecov](https://codecov.io/gh/codeclassroom/CodeRunner/branch/master/graph/badge.svg)](https://codecov.io/gh/codeclassroom/CodeRunner)
![PyPI - Format](https://img.shields.io/pypi/format/coderunner?color=orange)
![PyPI - Python Version](https://img.shields.io/pypi/pyversions/coderunner)
[![Documentation Status](https://readthedocs.org/projects/coderunner/badge/?version=latest)](https://coderunner.readthedocs.io/en/latest/?badge=latest)
![PyPI - Downloads](https://img.shields.io/pypi/dm/coderunner?color=blue)


## Installation

Install using `pip` from PyPI

```bash
pip install coderunner
```

or directly from GitHub if you cannot wait to test new features

```bash
pip install git+https://github.com/codeclassroom/CodeRunner.git
```

## Usage

```python

import coderunner

source_code = "path-to/test_python.py"
language = "Python"
expected_output = "path-to/output.txt"
standard_input = "path-to/input.txt"

# use this if you have a standard input to Program
r = coderunner.code(source_code, language, expected_output, standard_input)

# otherwise
r = coderunner.code(source_code, language, expected_output)

# you can also ignore both fields
r = coderunner.code(source_code, language)

# Use path=False if not using file paths
r = coderunner.code("Hello, World", language, "Hello, World", path=False)

r.run()
print(r.getOutput())
print(r.getError())
# See Documentation for more methods.
```

## Documentation

> [CodeRunner Documentation](https://coderunner.readthedocs.io/en/latest/)


## Development

##### Prerequisites
- Python 3.6+
- virtualenv

1. Create virtual environment.
```bash
virtualenv -p python3 venv && cd venv && source bin/activate
```
2. Clone the repository.
```bash
git https://github.com/codeclassroom/CodeRunner.git
```
3. Install Dependencies.
```bash
pip install -r requirements.txt
```
4. Run tests.
```bash
python tests.py
```
5. Lint the project with
```bash
flake8 coderunner --max-line-length=88 --ignore=F401
black --check --diff coderunner
```

## 📝 Changelog

See the [CHANGELOG.md](CHANGELOG.md) file for details.

## :fire: Powered By
**[Judge0 API](https://github.com/judge0/api) - Free, robust and scalable open-source online code execution system**

## Author

👥 **Bhupesh Varshney**

- Twitter: [@bhupeshimself](https://twitter.com/bhupeshimself)
- DEV: [bhupesh](https://dev.to/bhupesh)

[![forthebadge](https://forthebadge.com/images/badges/built-with-love.svg)](https://forthebadge.com)

## 📜 License

This project is licensed under the MIT License. See the [LICENSE](LICENSE) file for details.

## 👋 Contributing

Please read the [CONTRIBUTING](CONTRIBUTING.md) guidelines for the process of submitting pull requests to us.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [1.0] - June , 2020
⚠ Major Update (Breaking Changes)

### Added
- Add API key authorizations
- Custom API URL support

### Changed
- Due to Judge0 API going [freemium](https://github.com/judge0/api/issues/171), the default API URL, [https://api.judge0.com] is no longer available. To use the API signup for a plan on [RapidAPI](https://rapidapi.com/hermanzdosilovic/api/judge0/pricing)


## [0.8] - May 27, 2020

### Fix
- Fixed bug where compiling a Java program resulted in Compilation Error.


## [0.7] - Jan 19, 2020

### Changed
- `code()` class now ignores `output`. i.e you can just provide source code & language to run a program.


## [0.6] - Jan 5, 2020

### Added

- New optional argument, `number_of_runs`  in `run()` method, use this to specify no.of times you want to run the code. Default is set to 1.
- Ported NEW Languages. CodeRunner now supports all languages provided by Judge0.
- `setFlags(options)` for setting options for the compiler (i.e. compiler flags).
- `setArguments(arguments)` for setting Command line arguments for the program.

### Changed
- Minor internal improvemets.


## [0.5] - Dec 20, 2019

### Added

- New instance method - `run()`.
- `run()` is now used to run the code i.e the code is submitted to Judge0 api using this method.
- Support for *Bash 4.4*.

### Changed
- Renamed Class `Run` to `code` for easier usage.
- `getStatus()`, now only returns the status in comparison to earlier versions where it performed multiple tasks.
This is effect fixes [#2](https://github.com/codeclassroom/CodeRunner/issues/2).

### Removed
 - `requests` as a dependency, Network requests are now 50% faster.


## [0.4] - Nov 11, 2019

### Added

- `getSubmissionDate()`, `getExitCode` new methods.
- Official Documentation.

### Changed

- Class Run `init` - Now you can pass _source code_, _input_ and _output_ to program as strings (limited to file paths in prior versions).


## [0.3] - Nov 9, 2019

### Added

- Removed redundant imports
- Added Module/Class docstrings for documentation
- Formatted Code


## [0.2] - Oct 31, 2019

### Changed

- Fix import requests problem.


## [0.1] - Oct 30, 2019
- Initial Release

test.py
import unittest
from judge import run

class TestRun(unittest.TestCase):
	def test_run(self):
		program_name = "testfiles/" + "test_java.java"
		language = "Java"
		output = "testfiles/" + "output.txt"
		self.assertEqual(run(program_name, language, output),
                     "Accepted", "Something Wrong")

if __name__ == '__main__':
    unittest.main()


tests.py
import unittest
from coderunner import coderunner
from coderunner.coderunner import InvalidURL, ValueTooLargeError
# # test for Java program


# class TestRunA(unittest.TestCase):
#     def test_run(self):
#         source_code = "testfiles/" + "test_java.java"
#         language = "Java"
#         output = "testfiles/output/" + "output.txt"
#         r = coderunner.code(source_code, language, output)
#         r.run()
#         self.assertEqual(r.getStatus(), "Accepted", "Something Wrong")


# # test for Java program with input


# class TestRunB(unittest.TestCase):
#     def test_run(self):
#         source_code = "testfiles/" + "test_java_input.java"
#         language = "Java"
#         output = "testfiles/output/" + "output5.txt"
#         Input = "testfiles/input/" + "input4.txt"
#         r = coderunner.code(source_code, language, output, Input)
#         r.run()
#         self.assertEqual(r.getStatus(), "Accepted", "Something Wrong")


# # test for Python3 program


# class TestRunC(unittest.TestCase):
#     def test_run(self):
#         source_code = "testfiles/" + "test_python.py"
#         language = "Python3"
#         output = "testfiles/output/" + "output6.txt"
#         r = coderunner.code(source_code, language, output)
#         r.run()
#         self.assertEqual(r.getStatus(), "Accepted", "Something Wrong")


# # test for Python3 program with input


# class TestRunD(unittest.TestCase):
#     def test_run(self):
#         source_code = "testfiles/" + "test_python_input.py"
#         language = "Python3"
#         output = "testfiles/output/" + "output2.txt"
#         Input = "testfiles/input/" + "input.txt"
#         r = coderunner.code(source_code, language, output, Input)
#         r.run()
#         self.assertEqual(r.getStatus(), "Accepted", "Something Wrong")


# # test for C program


# class TestRunE(unittest.TestCase):
#     def test_run(self):
#         source_code = "testfiles/" + "test_c.c"
#         language = "C"
#         output = "testfiles/output/" + "output7.txt"
#         r = coderunner.code(source_code, language, output)
#         r.run()
#         self.assertEqual(r.getStatus(), "Accepted", "Something Wrong")


# # test for C program with input


# class TestRunF(unittest.TestCase):
#     def test_run(self):
#         source_code = "testfiles/" + "test_c_input.c"
#         language = "C"
#         output = "testfiles/output/" + "output3.txt"
#         Input = "testfiles/input/" + "input2.txt"
#         r = coderunner.code(source_code, language, output, Input)
#         r.run()
#         self.assertEqual(r.getStatus(), "Accepted", "Something Wrong")


# # test for C++ program


# class TestRunG(unittest.TestCase):
#     def test_run(self):
#         source_code = "testfiles/" + "test_c++.cpp"
#         language = "C++"
#         output = "testfiles/output/" + "output8.txt"
#         r = coderunner.code(source_code, language, output)
#         r.run()
#         self.assertEqual(r.getStatus(), "Accepted", "Something Wrong")


# # test for C++ program with input


# class TestRunH(unittest.TestCase):
#     def test_run(self):
#         source_code = "testfiles/" + "test_c++_input.cpp"
#         language = "C++"
#         output = "testfiles/output/" + "output4.txt"
#         Input = "testfiles/input/" + "input3.txt"
#         r = coderunner.code(source_code, language, output, Input)
#         r.run()
#         self.assertEqual(r.getStatus(), "Accepted", "Something Wrong")


# # test to check ValueError exception


# class TestRunI(unittest.TestCase):
#     def test_run(self):
#         with self.assertRaises(ValueError):
#             coderunner.code("Hello World", "fgbh", "Hello World", path=False)


# # test to check OSError exception


# class TestRunJ(unittest.TestCase):
#     def test_run(self):
#         with self.assertRaises(OSError):
#             coderunner.code("Hello World", "C++", "Hello World")


# # test to check OSError exception


# class TestRunK(unittest.TestCase):
#     def test_run(self):
#         with self.assertRaises(OSError):
#             coderunner.code("testfiles/" + "test_c++.cpp", "C++", "Hello World")


# # test to check OSError exception


# class TestRunL(unittest.TestCase):
#     def test_run(self):
#         with self.assertRaises(OSError):
#             coderunner.code(
#                 "testfiles/" + "test_c++.cpp",
#                 "C++",
#                 "testfiles/output/" + "output4.txt",
#                 "my_input",
#             )


# # test to check Wrong Answer Status


# class TestRunM(unittest.TestCase):
#     def test_run(self):
#         source_code = 'print("This will return Wrong Answer")'
#         language = "Python3"
#         output = "Wrong Answer"
#         r = coderunner.code(source_code, language, output, path=False)
#         r.run()
#         self.assertEqual(r.getStatus(), "Wrong Answer", "Something Wrong")


# test to check invalid API url
class TestRunO(unittest.TestCase):
    def test_run(self):
        source_code = 'print("This will return Wrong Answer")'
        language = "Python3"
        output = "Wrong Answer"
        r = coderunner.code(source_code, language, output, path=False)
        with self.assertRaises(InvalidURL):
            r.api(key="ABCHE", url="ghthth")


# test to check ValueTooLargeError for Flags
class TestRunP(unittest.TestCase):
    def test_run(self):
        source_code = 'print("This will return Wrong Answer")'
        language = "Python3"
        output = "Wrong Answer"
        r = coderunner.code(source_code, language, output, path=False)
        with self.assertRaises(ValueTooLargeError):
            r.setFlags("qw"*70)


# test to check ValueTooLargeError for Arguments
class TestRunQ(unittest.TestCase):
    def test_run(self):
        source_code = 'print("This will return Wrong Answer")'
        language = "Python3"
        output = "Wrong Answer"
        r = coderunner.code(source_code, language, output, path=False)
        with self.assertRaises(ValueTooLargeError):
            r.setArguments("qw"*70)


if __name__ == "__main__":
    unittest.main()


test_java.java
import static java.lang.Math.pow;

// program to print a given pattern

class Main {
	public static void main (String[] args) {
		int p;
		for (int i = 1; i <= 10; i++) {
			p = (int) (pow (i, 2));
			System.out.print (p + " ");
			if (i == 4 || i == 7 || i == 9) System.out.println ();
		}
	}
}

demo.py
from coderunner import coderunner
import pprint
import os

from dotenv import load_dotenv

load_dotenv()

source_code = "testfiles/" + "test_python_input.py"
language = "Python3"
output = "testfiles/output/" + "output2.txt"
Input = "testfiles/input/" + "input.txt"


API_KEY = os.environ["API_KEY"]

r = coderunner.code(source_code, language, output, Input)

# Necessary step to initialize API keys & URL
r.api(key=API_KEY)

# r2 = coderunner.code("print(\"yo\")", "Python3", "YO", path=False)

# # run the code
r.run()

print("Run r :")
print("Status : " + r.getStatus())
print("Output : " + r.getOutput())

# r2.run()

# print("Run r2 :")
# print("Status : " + r2.getStatus())

# # check if any error occured
# if r.getError() is not None:
#     pprint.pprint("Error : " + r.getError())
# else:
#     print("Standard Output : ")
#     pprint.pprint(r.getOutput())
# print("Execution Time : " + r.getTime())
# print("Memory : " + str(r.getMemory()))


judge.py
import requests
import time
import sys

# language IDs on judge0, see README.md
languages = {
	"C++" : 10,
	"Java" : 27,
	"Python": 34,
	"C": 4
}

api_params = {
    "number_of_runs": "1",
    "cpu_time_limit": "2",
    "cpu_extra_time": "0.5",
    "wall_time_limit": "5",
    "memory_limit": "128000",
    "stack_limit": "64000",
    "max_processes_and_or_threads": "30",
    "enable_per_process_and_thread_time_limit": 'false',
    "enable_per_process_and_thread_memory_limit": 'true',
    "max_file_size": "1024"
}

API_URL = "https://api.judge0.com/submissions/"

def readCode(program):
	with open(program, 'r') as myfile:
		data = myfile.read()
	return data


def readExpectedOutput(output_file):
	with open(output_file, 'r') as out:
		data = out.read()
	return data


def readStandardInput(output_file):
	with open(output_file, 'r') as out:
		data = out.read()
	return data


def readStatus(token):
	while True:
		req = requests.get(API_URL + token['token'])
		response = req.json()
		status = response['status']['description']
		if status != "Processing" and status != "In Queue":
			break
		print(status)

	if status == "Accepted":
		#print(f'Output : \n{response["stdout"]}')
		print(f'Time : {response["time"]}')
		print("Compile Success ✅")
		return status
	else:
		return response['status']['description']

def submit(program, language_id, *argv):
	if len(argv) == 2:
		stdout = argv[0]
		stdin = argv[1]
		api_params['stdin'] = stdin
	elif len(argv) == 1:
		stdout = argv[0]
	
	api_params['expected_output'] = stdout
	api_params['language_id'] = language_id
	api_params['source_code'] = program

	res = requests.post(API_URL, data=api_params)
	token = res.json()
	return token


def run(program, language, *argv):
	program = readCode(program)
	stdout = readExpectedOutput(argv[0])
	
	if len(argv) == 2:
		stdin = readStandardInput(argv[1])
		token = submit(program, languages[language], stdout, stdin)
		status = readStatus(token)
	elif len(argv) == 1:
		token = submit(program, languages[language], stdout)
		status = readStatus(token)
	return status


CodeRunner/run.py
import time
import sys
import requests

# language IDs on judge0, see README.md
languages = {
	"C++" : 10,
	"Java" : 27,
	"Python": 34,
	"C": 4
}

api_params = {
    "number_of_runs": "1",
    "cpu_time_limit": "2",
    "cpu_extra_time": "0.5",
    "wall_time_limit": "5",
    "memory_limit": "128000",
    "stack_limit": "64000",
    "max_processes_and_or_threads": "30",
    "enable_per_process_and_thread_time_limit": 'false',
    "enable_per_process_and_thread_memory_limit": 'true',
    "max_file_size": "1024"
}

API_URL = "https://api.judge0.com/submissions/"

class coderunner:

	def __init__(self, program_name: str, lang: str, output: str, inp: str = None):
		self.program_name = program_name
		self.lang = lang
		if lang not in languages:
			raise ValueError(f"{lang} is not a supported languige {languages.keys()}")
		
		self.output = output
		self.inp = inp
		self.language_id = languages[lang]

	def __readCode(self):
		"""
		Read Source Code & return as string
		"""
		with open(self.program_name, 'r') as myfile:
			data = myfile.read()
		return data


	def __readExpectedOutput(self):
		with open(self.output, 'r') as out:
			data = out.read()
		return data


	def __readStandardInput(self):
		with open(self.inp, 'r') as out:
			data = out.read()
		return data


	def __readStatus(self, token: str):
		"""
		Check Submission status
		"""
		while True:
			req = requests.get(API_URL + token['token'])
			self.__response = req.json()
			self.__memory = self.__response["memory"]
			self.__time = self.__response["time"]
			status = self.__response['status']['description']
			if status != "Processing" and status != "In Queue":
				break

		if status == "Accepted":
			self.__stdout = self.__response["stdout"]
			return status
		else:
			return self.__response['status']['description']

	def __submit(self):
		if self.inp != None:
			api_params['stdin'] = self.inp
		
		api_params['expected_output'] = self.output
		api_params['language_id'] = self.language_id
		api_params['source_code'] = self.program_name

		res = requests.post(API_URL, data=api_params)
		token = res.json()
		return token


	def getStandardOutput(self):
		return self.__stdout


	def getMemory(self):
		return self.__memory


	def getError(self):
		if self.__response['stderr'] != '':
			return self.__response['stderr']
		return None


	def getTime(self):
		return self.__time


	def run(self):
		self.program_name = self.__readCode()
		self.output = self.__readExpectedOutput()
		
		if self.inp != None:
			self.inp = self.__readStandardInput()
			token = self.__submit()
			status = self.__readStatus(token)
		else:
			token = self.__submit()
			status = self.__readStatus(token)
		return status



coderunner/coderunner.py
"""
coderunner.py
====================================
The core module of CodeRunner
"""
import json
import os
import urllib.parse
import urllib.request

# language IDs on judge0, see Documentation
languages = {
    "Assembly": 45,
    "Bash": 46,
    "Basic": 47,
    "C": 50,
    "C++": 54,
    "C#": 51,
    "Common Lisp": 55,
    "D": 56,
    "Elixir": 57,
    "Erlang": 58,
    "Executable": 44,
    "Fortran": 59,
    "Go": 60,
    "Haskell": 61,
    "Java": 62,
    "JavaScript": 63,
    "Lua": 64,
    "OCaml": 65,
    "Octave": 66,
    "Pascal": 67,
    "PHP": 68,
    "Plain Text": 43,
    "Prolog": 69,
    "Python2": 70,
    "Python3": 71,
    "Ruby": 72,
    "Rust": 73,
    "TypeScript": 74,
}

api_params = {
    "cpu_time_limit": "2",
    "cpu_extra_time": "0.5",
    "wall_time_limit": "5",
    "memory_limit": "128000",
    "stack_limit": "64000",
    "max_processes_and_or_threads": "30",
    "enable_per_process_and_thread_time_limit": "false",
    "enable_per_process_and_thread_memory_limit": "false",
    "max_file_size": "1024",
}

HEADERS = {"x-rapidapi-host": "judge0.p.rapidapi.com", "useQueryString": True}

FIELDS = "?fields=stdout,memory,time,status,stderr,exit_code,created_at"


class ValueTooLargeError(Exception):
    """Raised when the input value is too large"""


class InvalidURL(Exception):
    """Raise when api_url is invalid"""


class code:
    """
    Args:
        - Source Code
        - Language
        - Expected Output
        - Standard Input (Optional)
        - path (optional)
    """

    def __init__(
        self,
        source: str,
        lang: str,
        output: str = None,
        inp: str = None,
        path: bool = True,
    ):

        self.path = path
        if lang not in languages:
            raise ValueError(f"{lang} is not a supported language {languages.keys()}")
        self.lang = lang
        self.language_id = languages[lang]
        self.__response = None
        self.__memory = None
        self.__time = None
        self.__stdout = None
        self.languages = list(languages.keys())

        if self.path:
            if not os.path.exists(source):
                raise OSError(f"{source} is not a valid file path")
            self.source = source

            if output is not None and not os.path.exists(output):
                raise OSError(f"{output} is not a valid file path")
            self.output = output

            if inp is not None and not os.path.exists(inp):
                raise OSError(f"{inp} is not a valid file path")
            self.inp = inp
        self.source = source
        self.output = output
        self.inp = inp

    def __readCode(self):
        try:
            with open(self.source, "r") as program_file:
                data = program_file.read()
            return data
        except FileNotFoundError as e:
            raise e

    def __readExpectedOutput(self):
        try:
            with open(self.output, "r") as exp_out:
                data = exp_out.read()
            return data
        except FileNotFoundError as e:
            raise e

    def __readStandardInput(self):
        try:
            with open(self.inp, "r") as standard_input:
                data = standard_input.read()
            return data
        except FileNotFoundError as e:
            raise e

    def __readStatus(self, token: str):
        """
        Check Submission Status
        """
        while True:
            req = urllib.request.Request(
                f'{self.API_URL}submissions/{token["token"]}{FIELDS}', headers=HEADERS
            )
            with urllib.request.urlopen(req) as response:
                req = response.read()

            self.__response = json.loads(req.decode("utf-8"))
            self.__memory = self.__response["memory"]
            self.__time = self.__response["time"]
            status = self.__response["status"]["description"]
            if status not in ("Processing", "In Queue"):
                break

        if status == "Accepted":
            self.__stdout = self.__response["stdout"]
            return status
        return self.__response["status"]["description"]

    def __submit(self):
        if self.inp is not None:
            api_params["stdin"] = self.inp
        if self.output is not None:
            api_params["expected_output"] = self.output

        api_params["language_id"] = self.language_id
        api_params["source_code"] = self.source

        post_data = urllib.parse.urlencode(api_params).encode("ascii")
        req = urllib.request.Request(
            f"{self.API_URL}submissions/", post_data, headers=HEADERS
        )
        with urllib.request.urlopen(req) as response:
            req = response.read()
        token = json.loads(req.decode("utf-8"))

        return token

    def api(self, key: str, url: str = None):
        """Setup API url and key"""
        HEADERS["x-rapidapi-key"] = key
        self.API_KEY = key
        if url is None:
            self.API_URL = "https://judge0.p.rapidapi.com/"
        else:
            user_api_url = urllib.parse.urlparse(url)
            if user_api_url.scheme and user_api_url.netloc:
                self.API_URL = url
            else:
                raise InvalidURL("Invalid API URL")

    def getSubmissionDate(self):
        """Submission date/time of program"""
        return self.__response["created_at"]

    def getExitCode(self):
        """Exitcode of program (0 or 1)"""
        return self.__response["exit_code"]

    def getOutput(self):
        """Standard output of the program"""
        return self.__stdout

    def getMemory(self):
        """Memory used by the program"""
        return self.__memory

    def getError(self):
        """Error occured during execution of program"""
        if self.__response["stderr"] != "":
            return self.__response["stderr"]
        return None

    def getTime(self):
        """Execution time of program"""
        return self.__time

    def setFlags(self, options: str):
        """Options for the compiler (i.e. compiler flags)"""
        if len(options) > 128:
            raise ValueTooLargeError("Maximum 128 characters allowed")
        api_params["compiler_options"] = options

    def setArguments(self, arguments: str):
        """Command line arguments for the program"""
        if len(arguments) > 128:
            raise ValueTooLargeError("Maximum 128 characters allowed")
        api_params["command_line_arguments"] = arguments

    def run(self, number_of_runs: int = 1):
        """Submit the source code on judge0's server & return status"""
        api_params["number_of_runs"] = number_of_runs

        if os.path.exists(self.source):
            if self.path:
                if self.inp is not None:
                    self.inp = self.__readStandardInput()
                if self.output is not None:
                    self.output = self.__readExpectedOutput()
                self.source = self.__readCode()

        token = self.__submit()
        self.__token = token

    def getStatus(self):
        """Submission status"""
        status = self.__readStatus(self.__token)

        return status
